### DISCUSSION TITLE
Encrypt sensitive credentials in Channels

### DISCUSSION THREAD

**Discussion by fraser-nds** [nextgenhealthcare/connect#5452] [Feature/Enhancement Ideas]
In a NextGen Connect channel there are a few places where sensitive information is stored. Most notably in our use-case, the passwords used for basic auth in both the source and destination HTTP senders/listeners.

![image](https://user-images.githubusercontent.com/79858320/194101201-b76ce596-6684-414d-870e-4d819e8ab95d.png)
![image](https://user-images.githubusercontent.com/79858320/194101313-47333066-29fd-4d1b-a5e0-a1d77afc1272.png)

While it is good that the password is obscured and not shown in plain text, it is still all too simple for a bad actor to export the channel, where it lies in plain text in the XML. These credentials have the potential to be very sensitive and it only takes the malicious party to break into NGC for them to also get free access into lots of other services that channel communicates with.

It would be useful if the source at least hashed the password before storing it in in the DB.

Meanwhile for the destination this would need to be encrypted at rest, and then decrypted when it is required to be used in a destination.

Hope you consider this feature enhancement!

**Comment by jonbartels** [comment]
The configuration map exists to solve this use case - https://docs.nextgen.com/bundle/Mirth_User_Guide_41/page/connect/connect/topics/c_Configuration_MapConfiguration_Map_connect_ug.html

For your screenshots the HTTP auth would change to have `${username}` and `${password}`. Then in your config map add the username and password to the config map. When your channel exports it will only have the tokens `${username}` and `${password}`. At runtime Mirth Connect will read the maps and replace those tokens with the actual values.

In addition to security, this also helps write channels in DEV, test them in TEST, and deploy to PROD _without_ changing the channel. Each iteration can just reference the config map then the config map values change on each environment.

**Reply by jonbartels** [reply]
I think a good improvement would be to update the documentation to include an explanation of what data is and is not exported with a channel and that config map is one way to help with that. 

@fraser-nds a similar spot that might contain sensitive data are the message templates. A common workflow is to capture any random message to use it as a message template. If that message contains sensitive data like PHI or PII then the channel is exported, now that sensitive data has been exposed. This second use case is NOT handled by using config map. The only way to address this is to not use sensitive data for a template.

**Reply by fraser-nds** [reply]
Hi @jonbartels, thanks for the speedy response. 

We have looked into using the config map and the global map, however we felt these were actually a little less secure as they are both exposed in the UI. From the config map for example you can simply toggle the Show Values option to expose the sensitive information. The global map is accessible from the the dashboard. And both create events in the event log that expose the values.

**Reply by siddharthsharma9258** [reply]
I always liked reading this response from Nick from 8 years ago.
https://forums.mirthproject.io/forum/mirth-connect/development/11599-encrypt-password-for-database-in-mirth-property?p=78429#post78429
